src/plugins/dkommon/dkommon.cpp

```diff
@@ -103,193 +103,325 @@
+{
+    std::set<std::string> drivers_list;
+    vmi_lock_guard vmi(drakvuf);
 
-    if (list_entry_va == flink && flink == blink &&
-        std::find(d->processes_list.begin(), d->processes_list.end(), info->attached_proc_data.pid) == d->processes_list.end())
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .pid = 4,
+    );
+
+    addr_t list_head = 0;
+    if (VMI_SUCCESS != vmi_read_addr_ksym(vmi, "PsLoadedModuleList", &list_head))
     {
-        d->processes_list.push_back(info->attached_proc_data.pid);
-        print_hidden_process_information(drakvuf, info);
+        PRINT_DEBUG("dkommon:enumerate drivers: failed to read PsLoadedModuleList value\n");
+        return drivers_list;
     }
-    else
+
+    list_head -= plugin->offsets[LDR_DATA_TABLE_ENTRY_INLOADORDERLINKS];
+
+    addr_t entry = list_head;
+    do
     {
-        goto done;
-    }
+        ctx.addr = entry + plugin->offsets[LDR_DATA_TABLE_ENTRY_INLOADORDERLINKS] + plugin->offsets[LIST_ENTRY_FLINK];
+        if (VMI_SUCCESS != vmi_read_addr(vmi.vmi, &ctx, &entry))
+        {
+            PRINT_DEBUG("dkommon:enumerate drivers: failed to read next entry (VA 0x%lx)\n", ctx.addr);
+            return drivers_list;
+        }
 
-err:
-    PRINT_DEBUG("[DKOMmon] Error. Failed to read virtual address.\n");
+        ctx.addr = entry + plugin->offsets[LDR_DATA_TABLE_ENTRY_FULLDLLNAME];
+        auto name = drakvuf_read_unicode_common(vmi.vmi, &ctx);
+        if (name && name->contents)
+        {
+            auto drv = std::string(reinterpret_cast<char*>(name->contents));
+            drivers_list.insert(split_string(drv, "\\").back());
+            vmi_free_unicode_str(name);
+        }
+    } while (entry != list_head);
 
-done:
-    drakvuf_release_vmi(drakvuf);
+    return drivers_list;
+}
 
+static event_response_t notify_zero_page_write(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = static_cast<dkommon*>(info->trap->data);
+    fmt::print(plugin->format, "dkommon", drakvuf, info,
+        keyval("Message", fmt::Qstr("Zero Page Write"))
+    );
     return 0;
 }
 
-static event_response_t notify_zero_page_write(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+static void process_visitor(drakvuf_t drakvuf, addr_t process, void* pass_ctx)
 {
-    dkommon* d = static_cast<dkommon*>(info->trap->data);
+    auto plugin = static_cast<dkommon*>(pass_ctx);
 
-    fmt::print(d->format, "dkommon", drakvuf, info,
-        keyval("Message", fmt::Qstr("Zero Page Write"))
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .pid = 4,
+        .addr = process + plugin->offsets[EPROCESS_UNIQUE_PROCESS_ID]
     );
 
-    return 0;
+    addr_t pid = 0;
+    vmi_lock_guard vmi(drakvuf);
+    if (VMI_SUCCESS != vmi_read_addr(vmi, &ctx, &pid))
+    {
+        PRINT_DEBUG("[DKOMMON] Failed to read process pid\n");
+        return;
+    }
+    plugin->live_processes.insert(pid);
```

Comment:
Why not just use `drakvuf_get_process_pid`?

Reply:
That's how I've been getting process PID all the time in exploitmon but I'm ok to change.

Reply:
You probably want to switch that over too, no point in duplicating that code.